middle_end/simplify_primitives.ml

```diff
@@ -37,6 +37,39 @@ let phys_equal (approxs:A.t list) =
     | Some (s1, Some f1), Some (s2, Some f2) -> Symbol.equal s1 s2 && f1 = f2
     | _ -> false
 
+let is_known_to_be_some_kind_of_int (arg:A.descr) =
+  match arg with
+  | Value_int _ | Value_char _ | Value_constptr _ -> true
+  | _ -> false
+
+let is_known_to_be_some_kind_of_block (arg:A.descr) =
+  match arg with
+  | Value_block _ | Value_float _ | Value_float_array _ | A.Value_boxed_int _
+  | Value_closure _ | Value_string _ -> true
+  | _ -> false
+
+let rec structurally_different (arg1:A.t) (arg2:A.t) =
+  match arg1.descr, arg2.descr with
+  | Value_block (tag1, fields1), Value_block (tag2, fields2) ->
+      not (Tag.equal tag1 tag2) ||
```

Comment:
Indentation needs fixing for these two stanzas (also on line 69 too).
I would also recommend following the GNU convention of putting the `||` at the start of each line of a multi-line conditional; I think it's easier to follow.